src/main/java/spoon/support/StandardEnvironment.java

```diff
@@ -418,7 +420,7 @@ public void setInputClassLoader(ClassLoader aClassLoader) {
 				if (onlyFileURLs) {
 					List<String> classpath = new ArrayList<>();
 					for (URL url : urls) {
-						classpath.add(url.getPath());
+						classpath.add(URLDecoder.decode(url.getPath(), StandardCharsets.UTF_8));
```

Comment:
I am not sure if this is better or something like
```java
classpath.add(Path.of(url.toURI()).toAbsolutePath().toString());
```

Reply:
I am unconvinced. Now that you remind me: Path.of(URL) will throw an exception if the URL is not pointing to something that lies on a filesystem or if no FileSystemProvider supporting the URL's protocol is present. For example to create a Path instance that points to a file located inside a Jar file (not the Jar itself), there has to be a FileSystem instance for the Jar file. In the case of the "jar:" protocol, Path.of(URL) will make a disk access and try to read the Jar contentand fail if the file is not present or cannot be read. You can try this with `jshell`:

```java
jshell> Path.of(new URL("jar:file:/hello.jar!/hello.txt").toURI())
|  Exception java.nio.file.FileSystemNotFoundException
|        at ZipFileSystemProvider.getFileSystem (ZipFileSystemProvider.java:156)
|        at ZipFileSystemProvider.getPath (ZipFileSystemProvider.java:142)
|        at Path.of (Path.java:208)
|        at (#3515 
```

Reply:
I doubt the existing code can handle files inside JARs either? URL-Decoding would return `file:/hello.jar!/hello.txt` which likely isn't handled correctly. I'd need to have a look at it, but failing early instead of *maybe* failing later on sounds like a plus to me?

Reply:
Hmm, this assumes utf8 encoding. How does this work on Windows when there's a non-ASCII character in the filepath?

Reply:
@slarse Maybe someone with access to a windows box could test. In any case, URLs *should* use UTF-8. Qutoe taken from [W3.org](https://www.w3.org/TR/html40/appendix/notes.html#non-ascii-chars):

> **Note.** Some older user agents trivially process URIs in HTML using the bytes of the character encoding in which the document was received. Some older HTML documents rely on this practice and break when transcoded. User agents that want to handle these older documents should, on receiving a URI containing characters outside the legal set, first use the conversion based on UTF-8. Only if the resulting URI does not resolve should they try constructing a URI based on the bytes of the character encoding in which the document was received.

Reply:
I am not fully convinced, that JDT accepts file paths referring to files *inside a jar* or other weird things correctly, so I am not sure raw URL decoding really is a benefit here over stricter Path.of validation/conversion.

Either way, I tried it out @slarse and the URLDecode javadoc states that it only uses the charset for percent-encoded data. `new URL("öäüß")` is not stored percent encoded, so the string is returned unchanged in the default JVM windows charset. The UTF-8 setting has no relevance for it, but if the URL does contain other percent encoded data it could maybe end up with a String using *two* encodings.

Reply:
@xzel23 I did not know that, good to know what the spec says (although I doubt all file systems follow that, a filepath is not necessarily a URI).

@I-Al-Istannen Thanks for checking it out, that's 100% confusing :D